README.md
# flutter_bounce

[![pub_package](https://img.shields.io/pub/v/flutter_bounce)](https://pub.dev/packages/flutter_bounce)
[![git_license](https://img.shields.io/github/license/aloklearning/flutter_bounce)](https://github.com/aloklearning/flutter_bounce/blob/master/LICENSE)
[![build](https://img.shields.io/appveyor/build/gruntjs/grunt)](https://pub.dev/packages/flutter_bounce#-analysis-tab-)

- An efficient flutter package for an **on-tap bounce** animation on any flutter widgets 
- This package is an enhanced version of [bouncing_widget](https://pub.dev/packages/bouncing_widget).
- **Limitation** of the **bouncing_widget** package: 
    - It is not efficient for the widgets in Scroll Pages. We cannot scroll while having our fingers on the widget
    - The same issue would be found [here](https://github.com/ThomasEcalle/bouncing_widget/issues/1)


## Advantages of flutter_bounce

- Flutter bounce package will give you full freedom, to be used in on both the **static** and **dynamic like scrolling pages**.
- Not sticky on scrolling, efficient bounce effect every where
- With **Duration**, controll the bounce animation
- It doesn't let the widget stay pressed when you scroll through, if we use the normal bounce effect with simply using the `Flutter AnimationController` and `Tranform.scale()` Widget. 

## Disclaimer

- Widget `flutter_widget` is not the copy paste of `bouncing_widget`, it works differently. Without limitations, and the functionality works differently, based upon the `duration`, and not on the `scaleFactor`. 
- It doesn't have the `scaleFactor` at all, which is required in the widget
- The code written in such a way that supports the `flutter_bounce` functionality. However, I have taken motivation from the limitations of the `bouncing_widget`

## Usage

To use this plugin, add `flutter_bounce` as a [dependency in your pubspec.yaml file](https://flutter.dev/docs/development/packages-and-plugins/using-packages).

## Demo

This is how the `Bounce` Widget works when wrapped around the widget. The below shows `Bounce` being used in both the scenario:
1. Static Page with Duration set facility
2. Scroll Page with default Duration as 100 ms:
    - Horizontal Scroll
    - Vertical Scroll

![](assets/simple-bounce.gif)
![](assets/scroll-bounce.gif)

## Example

- Import the library

```dart
import 'package:flutter_bounce/flutter_bounce.dart';
```

- In order to use this package, we just need to wrap the content with the widget `Bounce`.
- `child`, `duration` and `onPressed` is mandatory in the widget. 
- However, `onPressed` and `duration` can have `null` in the widget. By default `duration` is set to `100 milliseconds`.
- **Please Note:** Always pass the duration in `milliseconds` for best result

```dart
Bounce(
  duration: Duration(milliseconds: 110),
  onPressed: (){ YOUR_FUNCTION },
  child: YOUR_WIDGET
)
```

## Acknowledgement

- I would like to thank [Levent Kantaroğlu](https://github.com/leventkantaroglu) for contributing in the recent **null safety** addition to the code, which will now help the community to use this package efficiently.
- I would like to thank [ThomasEcalle](https://github.com/ThomasEcalle) for the indirect motivation via his widget, to come up with an efficient work around, which will now be useful for our flutter community people












CHANGELOG.md
## [1.1.0] - TODO: 14th Jun, 2021.

* TODO: Added null-safety support to the project

## [1.0.8] - TODO: 19th Apr, 2020 [IST]

* TODO: Build analysis report submitted

## [1.0.7] - TODO: 19th Apr, 2020 [IST]

* TODO: Changes revoked in README file for the package

## [1.0.6] - TODO: 19th Apr, 2020 [IST]

* TODO: Changes in the overall README file for the package

## [1.0.5] - TODO: 19th Apr, 2020 [IST]

* TODO: Changes in the overall README file for the package

## [1.0.4] - TODO: 19th Apr, 2020 [IST]

* TODO: Added license to the project

## [1.0.3] - TODO: 18th Apr, 2020

* TODO: Format changed in README

## [1.0.2] - TODO: 18th Apr, 2020

* TODO: Description changes in the Pubspec file

## [1.0.1] - TODO: 18th Apr, 2020

* TODO: Made changes as per the health check from the pub.dev

## [1.0.0] - TODO: 18th Apr, 2020.

* TODO: Added bounce library to the project


example/lib/main.dart
import 'package:flutter/material.dart';
import './bounce_demo.dart';
import './bounce_scroll.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Bounce',
      theme: ThemeData(
        //primarySwatch: Colors.indigo[400],
        primaryColor: Colors.indigo[400],
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(title: 'Flutter Bounce Demo'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);
  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {

  @override
  Widget build(BuildContext context) {
    
    return Scaffold(
      appBar: AppBar(
        centerTitle: true,
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            RaisedButton(
              onPressed: (){ Navigator.push(context, new MaterialPageRoute(
                builder: (context) => SimpleBounceDemo())
              ); },
              elevation: 5.0,
              color: Theme.of(context).primaryColor,
              textColor: Colors.white,
              child: Text('Simple Bounce Demo')
            ),
            SizedBox(height: 20.0),
            RaisedButton(
              elevation: 5.0,
              onPressed: (){ Navigator.push(context, new MaterialPageRoute(
                builder: (context) => BounceScroll())); },
              textColor: Colors.white,
              color: Theme.of(context).primaryColor,
              child: Text('Bounce with Scroll')
            )
          ]
        )
       ) // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}


lib/flutter_bounce.dart
library flutter_bounce;

import 'package:flutter/material.dart';

class Bounce extends StatefulWidget {
  final VoidCallback onPressed;
  final Widget child;
  final Duration duration;
  // This will get the data from the pages
  // Makes sure child won't be passed as null
  Bounce(
      {required this.child, required this.duration, required this.onPressed});

  @override
  BounceState createState() => BounceState();
}

class BounceState extends State<Bounce> with SingleTickerProviderStateMixin {
  late double _scale;

  // This controller is responsible for the animation
  late AnimationController _animate;

  //Getting the VoidCallack onPressed passed by the user
  VoidCallback get onPressed => widget.onPressed;

  // This is a user defined duration, which will be responsible for
  // what kind of bounce he/she wants
  Duration get userDuration => widget.duration;

  @override
  void initState() {
    //defining the controller
    _animate = AnimationController(
        vsync: this,
        duration: const Duration(
            milliseconds: 200), //This is an inital controller duration
        lowerBound: 0.0,
        upperBound: 0.1)
      ..addListener(() {
        setState(() {});
      }); // Can do something in the listener, but not required
    super.initState();
  }

  @override
  void dispose() {
    // To dispose the contorller when not required
    _animate.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    _scale = 1 - _animate.value;
    return GestureDetector(
        onTap: _onTap,
        child: Transform.scale(
          scale: _scale,
          child: widget.child,
        ));
  }

  //This is where the animation works out for us
  // Both the animation happens in the same method,
  // but in a duration of time, and our callback is called here as well
  void _onTap() {
    //Firing the animation right away
    _animate.forward();

    //Now reversing the animation after the user defined duration
    Future.delayed(userDuration, () {
      _animate.reverse();

      //Calling the callback
      onPressed();
    });
  }
}


example/lib/bounce_demo.dart
import 'package:flutter/material.dart';
import 'package:flutter_bounce/flutter_bounce.dart';

class SimpleBounceDemo extends StatefulWidget {
  @override
  SimpleBounceDemoState createState() => SimpleBounceDemoState();
}

class SimpleBounceDemoState extends State<SimpleBounceDemo> {
  int _duration = 100;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Simple Bounce Demo', 
          style: TextStyle(color: Colors.white)),
        centerTitle: true
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            Bounce(
              onPressed: null,
              duration: Duration(milliseconds: _duration),
              child: Container(
                decoration: BoxDecoration(
                  color: Theme.of(context).primaryColor,
                  shape: BoxShape.circle,
                ),
                child: Padding(
                  padding: const EdgeInsets.all(12.0),
                  child: Icon(Icons.add, color: Colors.white)
                ),
              ),
            ),
            SizedBox(height: 30.0),
            Bounce(
              duration: Duration(milliseconds: _duration), 
              onPressed: null,
              child: Container(
                height: 50.0,
                width: MediaQuery.of(context).size.width * 0.5,
                decoration: BoxDecoration(
                  color: Theme.of(context).primaryColor,
                  borderRadius: BorderRadius.circular(80.0)
                ),
                child: Center(
                  child: Text('Click to bounce', 
                    style: TextStyle(color: Colors.white, fontSize: 18.0)
                ))
              )
            ),
            SizedBox(height: 20.0),
            Container(
              width: MediaQuery.of(context).size.width * .5,
              child: Slider(
                activeColor: Theme.of(context).primaryColor,
                min: 50,
                max: 300,
                value: _duration.toDouble(),
                onChanged: (newValue) {
                  setState(() {
                    _duration = newValue.toInt();
                  });
                }
              )
            ),
            Center(
              child: Text(
                "Duration = $_duration ms",
                style: TextStyle(
                  color: Theme.of(context).primaryColor,
                  fontSize: 18,
                )
              )
            )
          ]
        )
      )
    );
  }
}

example/lib/bounce_scroll.dart
import 'package:flutter/material.dart';
import 'package:flutter_bounce/flutter_bounce.dart';

class BounceScroll extends StatelessWidget {

  //This is the widget for showing up in scroll
  List<Widget> get verticalList{
    List<Widget> _widget = [SizedBox(height: 20.0)];

    for(int i=0; i<10; i++){
      _widget.add(verticalListCards);
      _widget.add(SizedBox(height: 10.0));
    }

    return _widget;
  }

  //This is the widget
  Widget get verticalListCards{
    return Bounce(
      duration: Duration(milliseconds: 110),
      onPressed: null,
      child: Container(
        height: 80.0,
        decoration: BoxDecoration(
          color: Colors.indigo[400],
          borderRadius: BorderRadius.circular(10.0)
        )
      )
    );
  }

  //This is for horizontal list
  List<Widget> get horizontalList{
    List<Widget> _view = [SizedBox(width: 20.0)];

    for(int i=0; i<10; i++){
      _view.add(horizontalListWidget);
      _view.add(SizedBox(width: 10.0));
    }

    return _view;
  }

  Widget get horizontalListWidget{
    return Bounce(
      duration: Duration(milliseconds: 110),
      onPressed: null,
      child: Container(
        height: 70.0,
        width: 200.0,
        decoration: BoxDecoration(
          color: Colors.indigo[400],
          borderRadius: BorderRadius.circular(10.0)
        )
      )
    );
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Bounce with Scroll Demo', 
          style: TextStyle(color: Colors.white)),
        centerTitle: true
      ),
      body: Column(
        mainAxisSize: MainAxisSize.min,
        crossAxisAlignment: CrossAxisAlignment.stretch,
        children: <Widget>[
          Container(
            padding: EdgeInsets.symmetric(horizontal: 15.0),
            height: MediaQuery.of(context).size.height * 0.5,
            child: ListView(
              shrinkWrap: true,
              scrollDirection: Axis.vertical,
              children: this.verticalList
            )
          ),
          Expanded(
            child: Container(
              padding: EdgeInsets.symmetric(vertical: 15.0),
              child: ListView(
                shrinkWrap: true,
                scrollDirection: Axis.horizontal,
                children: this.horizontalList
              )
            )
          )
        ]
      )
    );
  }
}